src/microbe_stage/ProcessSystem.cs

```diff
@@ -207,6 +212,77 @@ void MakeSureResultExists(Compound compound)
         return ComputeCompoundBalance(organelles.Select(o => o.Definition), biome, amountType);
     }
 
+    /// <summary>
+    ///   Computes the compound balances for given organelle list in a patch and at a given time during the day (or
+    ///   using longer timespan values)
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     Assumes that the cell produces at most as much ATP as it consumes
+    ///   </para>
+    /// </remarks>
+    public static Dictionary<Compound, CompoundBalance> ComputeCompoundBalanceAtEquilibrium(
+        IEnumerable<OrganelleTemplate> organelles, BiomeConditions biome, CompoundAmountType amountType,
+        EnergyBalanceInfo energyBalance)
+    {
+        var result = new Dictionary<Compound, CompoundBalance>();
+
+        void MakeSureResultExists(Compound compound)
+        {
+            if (!result.ContainsKey(compound))
+            {
+                result[compound] = new CompoundBalance();
+            }
+        }
+
+        float consumptionProductionRatio = energyBalance.TotalConsumption / energyBalance.TotalProduction;
+        bool useRatio;
+
+        foreach (var organelle in organelles)
+        {
+            foreach (var process in organelle.Definition.RunnableProcesses)
+            {
+                var speedAdjusted = CalculateProcessMaximumSpeed(process, biome, amountType);
+
+                useRatio = false;
+
+                // If the cell produces more ATP than it needs, its ATP producing processes need to be toned down
+                if (speedAdjusted.Outputs.Select(o => o.Key).Contains(ATP) && consumptionProductionRatio < 1.0f)
```

Comment:
```suggestion
                if (speedAdjusted.Outputs.ContainsKey(ATP) && consumptionProductionRatio < 1.0f)
```
Would that work?

Reply:
Unfortunately not, `Outputs` doesn't have the `ContainsKey` method or anything similar as far as I can tell.

Reply:
I think refactoring the outputs property to be like this: `public IReadOnlyDictionary<Compound, float> Outputs => WritableOutputs;` would be a good choice here.